Add UserOverrideableDKIMRegistry.sol.

wshino opened the pull request:
## Description

Relates to https://github.com/zkemail/ether-email-auth/pull/22

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I have discussed with the team prior to submitting this PR
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes

Coverage

```
| File                                                                               | % Lines          | % Statements     | % Branches      | % Funcs         |
| UserOverrideableDKIMRegistry.sol                                                   | 96.55% (56/58)   | 93.06% (67/72)   | 73.21% (41/56)  | 90.00% (9/10)   |
```

SoraSuegami commented:
Thank you for your PR!
Could you modify the logic as follows?
1. The contract also stores address mainAuthorizer, which is provided when the contract is initialized.

2. For every function, if dkimPublicKeyHashes[domainName][publicKeyHash][mainAuthorizer]  is true,  a variable setThreshold increases by 1. If dkimPublicKeyHashes[domainName][publicKeyHash][msg.sender]  is true,  a variable setThreshold increases by 2.
A variable revokeThreshold is defined in the same way for revokedDKIMPublicKeyHashes.

3. The isDKIMPublicKeyHashValid function returns false if 1) revokeThreshold is larger than or equal to 1 or 2) setThreshold is less than 2. Otherwise, it returns true. This spec implies that 1) either the user or the main authorizer such as the canister signer can revoke a public key alone, and 2) setting a new public key requires the user's signature even when the main authorizer already approves it.

4. The setDKIMPublicKeyHash function takes as input domainName, publicKeyHash, address authorizer, and bytes signature. First, it computes revokeThreshold for domainName and publicKeyHash and requires revokeThreshold < 2.  Second, if msg.sender == authorizer, it sets dkimPublicKeyHashes[domainName][publicKeyHash][authorizer] = true and ends the process. Otherwise, it verifies signature as an EIP1271 signature if authorizer is a contract address, and verifies signature as an ECDSA signature otherwise. If and only if the signature is valid, sets  dkimPublicKeyHashes[domainName][publicKeyHash][authorizer] = true.

5. The revokeDKIMPublicKeyHash function takes as input publicKeyHash, address authorizer, and bytes signature. First, it computes revokeThreshold for  publicKeyHash and requires revokeThreshold < 2.  Second, if msg.sender == authorizer, it sets revokedDKIMPublicKeyHashes[publicKeyHash][authorizer] = true and ends the process. Otherwise, it verifies signature as an EIP1271 signature if authorizer is a contract address, and verifies signature as an ECDSA signature otherwise. If and only if the signature is valid, sets  revokedDKIMPublicKeyHashes[publicKeyHash][authorizer] = true.

wshino commented on packages/contracts/UserOverrideableDKIMRegistry.sol:
```diff
@@ -0,0 +1,268 @@
+            } else {
+                address recoveredSigner = digest.recover(signature);
+                require(
+                    recoveredSigner == authorizer,
```
Isn't the recoveredSigner the mainAuthorizer, not the authorizer?

wshino commented:
@SoraSuegami 
1. I've just fixed some test cases and README.md.
2. Also I updated the natspec for public functions in UserOverrideableDKIMRegistry.sol.
3. For reference, I've pasted the coverage result of the contract.

SoraSuegami commented:
@saleel Do you think we can ignore the above socket security alert?

saleel commented:
@SocketSecurity ignore npm/@zk-email/helpers@6.1.1

saleel commented:
> @saleel Do you think we can ignore the above socket security alert?

I think its ok. The alert is because of the dependency being a http url, but its from a trusted repo.

SoraSuegami commented:
@SocketSecurity ignore npm/@zk-email/zk-regex-circom@2.1.0
